### PR TITLE
bpo-32124: Document C functions safe before init

### PR DESCRIPTION
**Pull request by vstinner**
Explicitly document C functions and C variables that can be set
before Py_Initialize().

https://bugs.python.org/issue32124

### CODE REVIEW

File: Doc/c-api/init.rst
Comment (Mariatta) [inline Doc/c-api/init.rst]: Should there be a `*` in the beginning, like the rest?

File: Doc/c-api/init.rst
Comment (serhiy-storchaka) [inline Doc/c-api/init.rst]: Is not this the *all* flags?

File: Doc/c-api/init.rst
Comment (serhiy-storchaka) [inline Doc/c-api/init.rst]: PyImport_AppendInittab(), PyImport_ExtendInittab(), PyInitFrozenExtensions().

File: Doc/c-api/init.rst
Comment (serhiy-storchaka) [inline Doc/c-api/init.rst]: If `Py_GetProgramName` is included, shouldn't other `Py_Get*` functions be included? Py_GetPath(), Py_GetVersion(), and many others.

File: Doc/c-api/init.rst
Comment (serhiy-storchaka) [inline Doc/c-api/init.rst]: Wrong indentation (should be 3 spaces).

Document what is the value if use `-bb`. And the same for `-OO` below.

File: Doc/c-api/init.rst
Comment (serhiy-storchaka) [inline Doc/c-api/init.rst]:
```
.. c:var:: Py_BytesWarningFlag
```

File: Doc/c-api/init.rst
Comment (vstinner) [reply Doc/c-api/init.rst]: Typo, fixed

File: Doc/c-api/init.rst
Comment (vstinner) [reply Doc/c-api/init.rst]: Done

File: Doc/c-api/init.rst
Comment (vstinner) [reply Doc/c-api/init.rst]: Done

File: Doc/c-api/init.rst
Comment (vstinner) [reply Doc/c-api/init.rst]: I changed the text.

File: Doc/c-api/init.rst
Comment (vstinner) [reply Doc/c-api/init.rst]: Py_GetPath() must not be called before Py_Initialize().

I added other Py_GetXXX() functions.

**Comment by vstinner** [comment]
@Mariatta, @serhiy-storchaka: I updated my PR. Would you mind to review it again?

File: Doc/c-api/init.rst
Comment (vstinner) [inline Doc/c-api/init.rst]: I'm not sure that it's a good idea to document this flag, since I proposed to remove the flag https://bugs.python.org/issue32125

File: Doc/c-api/init.rst
Comment (Mariatta) [inline Doc/c-api/init.rst]: Typo :) `safetely` -> `safely`

File: Doc/c-api/init.rst
Comment (serhiy-storchaka) [inline Doc/c-api/init.rst]:
```
See :pep:`529` for more details.
```

File: Doc/c-api/init.rst
Comment (serhiy-storchaka) [inline Doc/c-api/init.rst]: I would group Py_GetBuildInfo() etc as "informative functions".

File: Doc/c-api/init.rst
Comment (Mariatta) [inline Doc/c-api/init.rst]: I see the word **wizards** and thought of magicians :) Do you mean something like setup wizard?
Or only a real wizard/magician can use this flag? 😅 
Maybe this is a common terminology and I'm just too new.. 🤷‍♀️

File: Doc/c-api/init.rst
Comment (Mariatta) [inline Doc/c-api/init.rst]:
```
See also :pep:`529`.
```
 (without "the")

File: Doc/c-api/init.rst
Comment (Mariatta) [inline Doc/c-api/init.rst]: Remove "the".

**Review by Mariatta — CHANGES_REQUESTED** [review]
Noticed a typo and grammar issues :)

File: Doc/c-api/init.rst
Comment (serhiy-storchaka) [reply Doc/c-api/init.rst]: I think Victor meant real [wizards](http://www.catb.org/jargon/html/W/wizard.html), like his and me. :tophat:

File: Doc/c-api/init.rst
Comment (serhiy-storchaka) [inline Doc/c-api/init.rst]: Isn't the third colon redundant?

**Review by serhiy-storchaka — COMMENTED** [review]
What about adding an explicit note in the description of all of these functions that they *can* be called before Py_Initialize()?

File: Doc/c-api/init.rst
Comment (Mariatta) [reply Doc/c-api/init.rst]: Ok in that case I suggest something less ambiguous, like "for expert" or "for experienced core devs" only.
This sentence also missed a period (`.`) at the end.

Comment (serhiy-storchaka) [reply Doc/c-api/init.rst]: This wording already is used two times in the documentation. In particularly in the documentation for the `-d` option. This is very old documentation.

Comment (serhiy-storchaka) [reply Doc/c-api/init.rst]: In sum, this option is only for those who understand this jargon. :wink:

File: Doc/c-api/init.rst
Comment (vstinner) [reply Doc/c-api/init.rst]: Fixed.

Crap. "safetely" looked weird, so I used my spell check: "Google safetely", but I found results. Maybe there is an issue with my spell checker? :-)

File: Doc/c-api/init.rst
Comment (vstinner) [reply Doc/c-api/init.rst]: I copied the doc from the -d option documentation.

Ok, I replaced "for wizards only" with "for expert only".

File: Doc/c-api/init.rst
Comment (vstinner) [reply Doc/c-api/init.rst]: I picked: See :pep:`529` for more details.

File: Doc/c-api/init.rst
Comment (vstinner) [reply Doc/c-api/init.rst]: Removed

File: Doc/c-api/init.rst
Comment (vstinner) [reply Doc/c-api/init.rst]: Oops, it's a mistake. Fixed.

**Comment by vstinner** [comment]
I have made the requested changes; please review again.

File: Doc/c-api/init.rst
Comment (serhiy-storchaka) [reply Doc/c-api/init.rst]: I would keep the original wording. It is used also in the man page since the initial edition in 1994.

**Review by Mariatta — APPROVED** [review]

**Comment by Mariatta** [comment]
Thanks @vstinner.
I personally prefer the word "experts" over "wizards". It's not a jargon or idiom, something the general public will understand :)
Thanks for changing it.

File: Doc/c-api/init.rst
Comment (ericsnowcurrently) [inline Doc/c-api/init.rst:59]: Would you mind putting these into a list?

**Comment by serhiy-storchaka** [comment]
This option is not for the general public.

If change the wording, you should change also the man page. I think it is better to open a separate issue for changing wizards to experts. This change is not related to this issue.

File: Doc/c-api/init.rst
Comment (ericsnowcurrently) [inline Doc/c-api/init.rst:26]: Why is PyInitFrozenExtensions() in this list?

**Comment by vstinner** [comment]
> This option is not for the general public.

"expert only" doesn't mean "generic public", it's more the opposite.

> If change the wording, you should change also the man page.

Done.

File: Doc/c-api/init.rst
Comment (ericsnowcurrently) [inline Doc/c-api/init.rst:29]: Why are we promising that PyObject_SetArenaAllocator() (and PyObject_GetArenaAllocator) is available before initialization?  What do embedders need it for before init?  Can we hold off adding it to this list until necessary?

**Comment by vstinner** [comment]
Thank you for your reviews @serhiy-storchaka and @Mariatta! I'm not confortable when I have to modify the documentation, so your feedback made me more confident in my change :-)

File: Doc/c-api/init.rst
Comment (ericsnowcurrently) [inline Doc/c-api/init.rst:27]: How much of the PyMem_*() API do we really need to promise to embedders?  AFAICS, we've only required them to use PyMem_RawFree() before initialization, due to Py_DecodeLocale().  The less we promise pre-init, the better.


Ideally we would provide equivalent pre-init functions, along with pre-init allocator state, that embedders could use if they needed.
If we do end up needing more, could we instead provide equivalent pre-init functions that take explicit allocator state that

File: Doc/c-api/init.rst
Comment (ericsnowcurrently) [inline Doc/c-api/init.rst:67]: In some ways, I'd rather we not advertise these flags as global variables, and particularly not as stable pre-init state.  If the goal is PEP 432 then further publicizing this state helps to entrench and persist the API from which we're trying to move away.

On the other hand, PEP 432 and its API is not accepted (yet) and the status quo suggests that the documentation you've added here for existing state is useful.

From my biased position, I'd favor leaving this part out for now, but only if we're serious about PEP 432. :)

**Comment by ericsnowcurrently** [comment]
Sorry I didn't get my review out sooner.

File: Doc/c-api/init.rst
Comment (ncoghlan) [reply Doc/c-api/init.rst:67]: For PEP 432, I want us to consider "Don't break backwards compatibility with the current embedding API" as a hard constraint. Part of that process is going to be reverse engineering what the current embedding API actually *is* in practice, since we don't have it fully documented, and we definitely don't have it fully tested.

So in a post-PEP-432 world, we'll end up with a situation where we have a "Legacy Embedding API" that we promise not to break (at least, not prior to Python 4.0 at the very earliest), and then a more fine-grained configuration-struct-based one.

File: Doc/c-api/init.rst
Comment (ncoghlan) [reply Doc/c-api/init.rst:27]: Setting the allocators pre-initialization is required, as that allows the embedding application to control exactly how CPython allocates memory during startup.

File: Doc/c-api/init.rst
Comment (ncoghlan) [reply Doc/c-api/init.rst:29]: Same reason as for `PyMem_SetAllocators`: the embedding application needs to be able to specify how memory allocation during `Py_Initialize()` itself should be handled.

**Review by ncoghlan — COMMENTED** [review]
Thanks @vstinner, this looks good to me, and defines a solid set of backwards compatibility constraints for the PEP 432 refactoring work.

File: Doc/c-api/init.rst
Comment (vstinner) [reply Doc/c-api/init.rst:26]: It was a request from @serhiy-storchaka. I don't know the rationale :-)

File: Doc/c-api/init.rst
Comment (vstinner) [reply Doc/c-api/init.rst:67]: > In some ways, I'd rather we not advertise these flags as global variables, and particularly not as stable pre-init state.

What do you mean by "not stable"? The inital state of these flags is well defined.

These flags are already actively used in the wild, it's lilely to only way to tune some Python options.

Not documenting them only makes Python harder to embed, it doesn't prevent users to use it.

Moreover, variables are exported as public symbols (prefixed with "Py", not "_Py").

**Comment by vstinner** [comment]
@ericsnowcurrently: I created the PR #4621 to change the formatting of the note. I also replied to your comments. Do you think that someone should be changed in the new PR #4621? Maybe comment there directly.

File: Doc/c-api/init.rst
Comment (serhiy-storchaka) [reply Doc/c-api/init.rst:26]: It is called before `Py_Initialize()` in `Python/frozenmain.c` on Windows.

**Pull request merged**
